docs/proposals/Periodic-Retrieval-of-Certificates-and-Keys-from-AKV.md

```diff
@@ -0,0 +1,121 @@
+# Periodic retrieval of certificates and keys from Azure Key Vault
+
+## Problem/Motivation
+
+When ensuring image integrity and authenticity, you can sign them using either Notation with certificates in Azure Key Vault (AKV) or Cosign with key pairs stored in AKV. To verify these signed images with Ratify in K8s, users typically configure the corresponding certificates or keys using custom resource `KeyManagementProvider`. This allows Ratify to retrieve the correct certificates or keys for verification.
+
+As a security best practice, certificates and keys are rotated regularly. This rotation can occur automatically, resulting in the generation of new versions of the existing certificates or keys in AKV. A certificate may also be revoked due to reasons such as a compromised private key or an invalid certificate identity. In such cases, the certificate is rotated on demand to a new version. Similarly, keys in AKV may be disabled for security reasons, prompting users to rotate them with a new version timely.
+
+Regardless of whether it’s certificate or key rotation, a new version is typically generated for the existing certificate or key. In the current Ratify version (v1.2.0 or prior), when rotation occurs, users must manually apply the `KeyManagementProvider` resource to trigger Ratify to retrieve the latest/specific versions of certificates or keys. This manual step can be cumbersome for users, especially considering automated certificate and key rotation. Users also need to keep the old versions of certificates or keys configured for some periods, as not all images are signed with the latest/specific versions in a short period. In practice, it’s challenging to keep up with these changes manually, and misconfigurations can lead to image verification failures and thus unnecessary service downtime. To address this, automated processes for periodically retrieving all available versions of certificates or keys configured in `KeyManagementProvider` resources would greatly benefit users.
+
+This proposal only covers the case of using AKV as the key management provider. Other providers may have different scenarios that need separate proposals. This proposal does not address the case of verifying images that were signed with Notary Project signatures before the certificate expired, and Ratify handles this case with timestamp support.
+
+## Scenarios
```

Comment:
Would like discuss error scenarios too. On non system errors, e.g. disabled key , identity permission error, are we safe to evict cached certs. Are there any non system errors we want to keep the cached certs?

Reply:
I am not sure about the errors you referred to. If the key is disabled, it should be removed from the cache in my opinion, otherwise, it can still be used for verification, right? Or from design point of view, you want to keep the cache, but mark it as disabled? It's better you can elaborate your error scenarios a bit.

Reply:
We ll set a default state to persist the cache since the set of possible errors are large. We can keep a list of scenarios key/cert will be evicted. 
- All disabled keys are removed
- Keys removed on 4XX response including 404 (not found)/403 ( unauthorized)